pkg/controller/metrics/metrics_with_dynamic_labels.go

```diff
@@ -0,0 +1,155 @@
+package metrics
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
+	log "github.com/sirupsen/logrus"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"sigs.k8s.io/controller-runtime/pkg/metrics"
+
+	hivev1 "github.com/openshift/hive/apis/hive/v1"
+	"github.com/openshift/hive/apis/hive/v1/metricsconfig"
+	"github.com/openshift/hive/pkg/constants"
+)
+
+// metricsWithDynamicLabels encompasses metrics that can have optional labels.
+type metricsWithDynamicLabels interface {
+	Observe(metav1.Object, map[string]string, float64)
+	Register()
+}
+
+type dynamicLabels struct {
+	// fixedLabels defines the label keys that are always registered on the metric.
+	fixedLabels []string
+	// optionalLabels is generated dynamically based on HiveConfig.Spec.MetricsConfig.MetricsWithDynamicLabels, and
+	// represents a map of metric label keys to ClusterDeployment metadata.label keys from which values will be gleaned
+	// when the metric is observed.
+	optionalLabels map[string]string
+}
+
+// getLabelList combines the fixed and optional labels and returns the list with just the labels that can be used to
+// define the metric
+func (d *dynamicLabels) getLabelList() []string {
+	var finalList []string
+	finalList = append(finalList, d.fixedLabels...)
+	for label := range d.optionalLabels {
+		finalList = append(finalList, label)
+	}
+	return finalList
+}
+
+// getRepeatedLabel detects and returns the first optional label it encounters, that overlaps with another label
```

Comment:
nit: This would be more useful if it returned _all_ repeated labels so the consumer doesn't have to iterate. Can be done in a fup though.